### DISCUSSION TITLE
http dialect+dbapi

### DISCUSSION THREAD

**Discussion by dboeckenhoff** [sqlalchemy/sqlalchemy#9829] [Usage Questions]
Hi. thanks for the nice project.

Background:

At our institute, due to extreme conditions of the experiment apparently data could not be stored in a usual database (I was told size and speed limitations). So our IT developed a so called "archive" and unfortunately a second data storage was established (so called "logbook", storing metadata on our experiments). This makes data retrieval inefficient since I have to query the logbook to select experiments based on their metadata and then only retrieve the data from the archive.

Question:

Does it make sense to implement a new dialect and dbapi for http request (bundling both servers) via requests/httpx? Based on the fields of the schema, the dbapi could differentiate which url to draw data from (archive or logbook). I think it would be beneficial to have the sqlalchemy frontend to the whole thing but have already dived in a little and see I have to learn a lot still, so before I fully commit to this Idea, I wanted to get the pro's opinions.

I know this is a little violation of how sqlalchemy is being inteded to be used but these boundary conditions are not set by me.

Thanks

**Comment by CaselIT** [comment]
Hi,

I'm not sure I follow what you are trying to do. Can you provide more information?

If you have some tables on a db while others are in another and you know which table is on which db you could maybe use something like this: https://docs.sqlalchemy.org/en/20/orm/persistence_techniques.html#session-partitioning

**Comment by zzzeek** [comment]
there's a standard pattern for this type of problem, where the live data that's written from some software has to be in some kind of non-standard format, massive / high speed experimental results are certainly a common example.

Basically you build a separate analytical / reporting database where you extract data from your high-speed experiment backend into a traditional RDBMS and that's where you do your SQL queries.

this addresses both problems:   1. speed - for the live capture of data, speed is maintained since you are using the specialized system   2. size - when you extract data from the experimental system into the traditional db, you can also take advantage of data transformations that aggregate across fine grained raw records that you don't need for your offline analysis.

**Comment by dboeckenhoff** [comment]
These are all very informative Points. This brings me way forward already. Thanks!

To illustrate my idea further, I made this little flowchart:

[![](https://mermaid.ink/img/eyJjb2RlIjoiZ3JhcGggVERcbkFbVXNlcl0gLS0-fFNRTCBxdWVyeXwgQ3tCYWNrZW5kIFN3aXRjaH1cbkMgLS0-fERhdGEgaXMgY2FjaGVkfCBEWyhUcmFkaXRpb25hbCBEQiBhcyBjYWNoZSldXG5EIC0tPiB8cmV0dXJufCBDXG5DIC0tPiB8cmV0dXJufCBBXG5DIC0tPnxEYXRhIG5vdCBjYWNoZWQsIGZvcndhcmQgcXVlcnl8IEV7SHR0cERpYWxlY3Q6SHR0cERiYXBpfVxuRSAtLT4gfHJlcXVlc3RzIGRhdGF8IEZbKGxvZ2Jvb2spXVxuRiAtLT4gfHJldHVybnwgRVxuRSAtLT4gfHJlcXVlc3RzIGRhdGF8IEdbKGFyY2hpdmUpXVxuRyAtLT4gfHJldHVybnwgRVxuRSAtLT4gfGNhY2hlfCBEICIsIm1lcm1haWQiOnsidGhlbWUiOiJkZWZhdWx0In0sInVwZGF0ZUVkaXRvciI6ZmFsc2V9)](https://mermaid-js.github.io/docs/mermaid-live-editor-beta/#/edit/eyJjb2RlIjoiZ3JhcGggVERcbkFbVXNlcl0gLS0-fFNRTCBxdWVyeXwgQ3tCYWNrZW5kIFN3aXRjaH1cbkMgLS0-fERhdGEgaXMgY2FjaGVkfCBEWyhUcmFkaXRpb25hbCBEQiBhcyBjYWNoZSldXG5EIC0tPiB8cmV0dXJufCBDXG5DIC0tPiB8cmV0dXJufCBBXG5DIC0tPnxEYXRhIG5vdCBjYWNoZWQsIGZvcndhcmQgcXVlcnl8IEV7SHR0cERpYWxlY3Q6SHR0cERiYXBpfVxuRSAtLT4gfHJlcXVlc3RzIGRhdGF8IEZbKGxvZ2Jvb2spXVxuRiAtLT4gfHJldHVybnwgRVxuRSAtLT4gfHJlcXVlc3RzIGRhdGF8IEdbKGFyY2hpdmUpXVxuRyAtLT4gfHJldHVybnwgRVxuRSAtLT4gfGNhY2hlfCBEICIsIm1lcm1haWQiOnsidGhlbWUiOiJkZWZhdWx0In0sInVwZGF0ZUVkaXRvciI6ZmFsc2V9)

I think this has elements of  @zzzeek proposal, does it? My Idea was to reuse some of the frontend to make the requests to logbook/archive more efficient. This way I can even avoid some http requests or reduce the size of the transfer, if some filtering on shots by logbook taggs tells me that I dont need to retrieve the large data of some shots from the archive. If I understand @zzzeek correctly, the usual approach would be to download ALL data from the relevant columns first into the intermediate traditional db, right?

Thanks fagain!

**Reply by CaselIT** [reply]
Not sure if it may make sense to handle all this on the db side, instead of the application side?
Sql db have a concept of foreign table, that may be useful here. I know postgresql has this feature https://www.postgresql.org/docs/current/postgres-fdw.html, but it should be available also in other dbs

**Reply by dboeckenhoff** [reply]
Again a super helpful hint. Thanks. I found https://wiki.postgresql.org/wiki/Foreign_data_wrappers#generic_web_wrappers and was especially cought by **https://github.com/pramsey/pgsql-http**. Unfortunately the docs are not very beginner friendly and I am not very deepn into this. Could you have a glance at this @CaselIT and give some comments/advice?

**Reply by CaselIT** [reply]
I've also never used this feature or pgsql-http, so can't offer a lot of advices.
It seems that pgsql-http is a web client, so if the external resource expects sql queries may not be the one to use.

Overall it seems that the syntax on that extension should be usable from sqlalchemy using `func` and/or `table()`
